### PR TITLE
build: add x86_64-w64-mingw toolchain

### PR DESCRIPTION
**Pull request by tamird**
Update to go1.8.1 while I'm here.

### CODE REVIEW

**Comment by BramGruneir** [comment]
LGTM, but perhaps someone more familiar with how this works should give it a quick scan.

File: build/Dockerfile [hunk 1]
```diff
@@ -71,23 +69,23 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
 # TODO(tamird): use the shallow clone when crosstool-ng 1.23 is released.
 # RUN git clone --depth 1 --branch crosstool-ng-1.23.0 https://github.com/crosstool-ng/crosstool-ng \
-RUN git clone https://github.com/crosstool-ng/crosstool-ng && git -C crosstool-ng checkout 1ec5e7aa3400b534264ba1c2fa3949dc6a2cca90 \
+RUN git clone https://github.com/crosstool-ng/crosstool-ng && git -C crosstool-ng checkout ee7662fa58a7f45ccd3e3520919ee70301ab7648 \
  && cd crosstool-ng \
  && ./bootstrap \
  && ./configure --prefix /usr/local/ct-ng \
  && make -j$(nproc) \
  && make install \
+ && cp ct-ng.comp /etc/bash_completion.d/ \
```
Comment (BramGruneir) [inline build/Dockerfile]: Bash completion?

File: build/Dockerfile [hunk 2]
```diff
@@ -116,8 +114,8 @@ ENV CXX /x-tools/x86_64-unknown-linux-gnu/bin/x86_64-unknown-linux-gnu-g++
 # BEGIN https://github.com/docker-library/golang/blob/master/1.8/alpine/Dockerfile
 
 COPY parallelbuilds-go1.8.patch /
-RUN curl -fsSL https://golang.org/dl/go1.8.src.tar.gz -o golang.tar.gz \
- && echo '406865f587b44be7092f206d73fc1de252600b79b3cacc587b74b5ef5c623596  golang.tar.gz' | sha256sum -c - \
+RUN curl -fsSL https://storage.googleapis.com/golang/go1.8.1.src.tar.gz -o golang.tar.gz \
```
Comment (BramGruneir) [inline build/Dockerfile]: Why the move away form golang.org/dl?

Comment (tamird) [inline build/Dockerfile]: I don't know, ask the go team. All the links on https://golang.org/dl/ point here now.

File: build/Dockerfile (hunk 1 above)
Comment (tamird) [inline build/Dockerfile]: Yes.

**Comment by benesch** [comment]
:lgtm:

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14791#-:-KhStokW_ir1-cJ417FY:bnfp4nl)*

**Comment by benesch** [comment]
Mentioning crosstool-ng/crosstool-ng#676 (which I assume is the reason for the crosstool-ng update, holler if it's not) for posterity.

File: build/Dockerfile [hunk 3]
```diff
@@ -71,23 +69,23 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
 
-COPY x86-linux-gnu.defconfig x86-linux-musl.defconfig ./
-RUN mkdir build && cd build \
- && rm -rf * && cp ../x86-linux-gnu.defconfig  defconfig && /usr/local/ct-ng/bin/ct-ng defconfig && CC=gcc CXX=g++ /usr/local/ct-ng/bin/ct-ng build \
- && rm -rf * && cp ../x86-linux-musl.defconfig defconfig && /usr/local/ct-ng/bin/ct-ng defconfig && CC=gcc CXX=g++ /usr/local/ct-ng/bin/ct-ng build \
+COPY x86_64-unknown-linux-gnu.defconfig x86_64-unknown-linux-musl.defconfig x86_64-w64-mingw.defconfig ./
+RUN mkdir build src && cd build \
+ && rm -rf * && cp ../x86_64-unknown-linux-gnu.defconfig  defconfig && /usr/local/ct-ng/bin/ct-ng defconfig && /usr/local/ct-ng/bin/ct-ng build \
+ && rm -rf * && cp ../x86_64-unknown-linux-musl.defconfig defconfig && /usr/local/ct-ng/bin/ct-ng defconfig && /usr/local/ct-ng/bin/ct-ng build \
+ && rm -rf * && cp ../x86_64-w64-mingw.defconfig          defconfig && /usr/local/ct-ng/bin/ct-ng defconfig && /usr/local/ct-ng/bin/ct-ng build \
```
Comment (jordanlewis) [inline build/Dockerfile]: I feel dumb - what is happening here? Why are you copying stuff to `.` and deleting it over and over again? Can you add a comment at least?

File: build/x86_64-w64-mingw.defconfig [hunk 4]
```diff
@@ -0,0 +1,19 @@
+CT_EXPERIMENTAL=y
+CT_ALLOW_BUILD_AS_ROOT=y
+CT_ALLOW_BUILD_AS_ROOT_SURE=y
```
Comment (jordanlewis) [inline build/x86_64-w64-mingw.defconfig]: lol at this double protection env variable, if that's what it is

File: build/Dockerfile [hunk 5]
```diff
@@ -50,7 +49,6 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
     cmake \
     file \
     flex \
-    fontconfig \
```
Comment (jordanlewis) [inline build/Dockerfile]: yay

**Review by jordanlewis — APPROVED** [review]
Science. Dog.

File: build/x86_64-w64-mingw.defconfig (hunk 4 above)
Comment (tamird) [inline build/x86_64-w64-mingw.defconfig]: It's not a env variable, but yes, always double bag it.

File: build/Dockerfile (hunk 3 above)
Comment (tamird) [inline build/Dockerfile]: I don't think there's a comment necessary here - `build` is the directory in which the build takes place. We have 3 configs that have meaningful names, but we have to give them the expected name before we can use them.

**Comment by tamird** [comment]
@benesch yes, that's the reason!

File: build/Dockerfile (hunk 3 above)
Comment (benesch) [inline build/Dockerfile]: @jordanlewis does raise the good point that you could instead run

```bash
rm -rf * && DEFCONFIG=../x86_64-blah.defconfig /usr/local/ct-ng/bin/ct-ng defconfig && /usr/local/ct-ng/bin/ct-ng build
```

if the man page isn't lying.

Comment (tamird) [inline build/Dockerfile]: Ah, seems to work. Rebuilding the image now (though it's probably unchanged).

**Comment by tamird** [comment]
TFYRs!

**Pull request merged**